File: redact-java/src/main/java/reuse/redact/Redact.java
package reuse.redact;

import java.util.function.Function;


public class Redact {
   protected final String pii;

   public Redact(String pii) {
      if (pii == null)
         throw new NullPointerException("input must not be null");

      this.pii = pii;
   }

   @Override
   public String toString() {
      return "<redacted " + strategy().apply(pii) + '>';
   }

   protected Function<String, String> strategy() {
      return Strategy.FULL;
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      Redact redact = (Redact) o;

      return pii.equals(redact.pii);
   }

   @Override
   public int hashCode() {
      return pii.hashCode();
   }
}

File: redact-kotlin/src/main/java/reuse/redact/Redact.kt
package reuse.redact

import java.util.function.Function

open class Redact(pii: String) {
    @JvmField
    protected val pii: String

    override fun toString(): String {
        return "<redacted " + redact().apply(pii) + '>'
    }

    protected open fun redact(): Function<String, String> {
        return Strategy.FULL
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val redact = o as Redact
        return pii == redact.pii
    }

    override fun hashCode(): Int {
        return pii.hashCode()
    }

    init {
        if (pii == null) throw NullPointerException("input must not be null")
        this.pii = pii
    }
}